File: dotnet/azure/Mcma.Azure.CosmosDb/CosmosDbMutex.cs
using System;
using System.Net;
using System.Threading.Tasks;
using Mcma.Data;
using Mcma.Logging;
using Microsoft.Azure.Cosmos;
using Newtonsoft.Json.Linq;

namespace Mcma.Azure.CosmosDb
{
    public class CosmosDbMutex : DocumentDatabaseMutex
    {
        public CosmosDbMutex(Container container, string partitionKeyName, string mutexName, string mutexHolder, TimeSpan? lockTimeout, ILogger logger = null)
            : base(mutexName, mutexHolder, lockTimeout, logger)
        {
            Container = container;
            PartitionKeyName = partitionKeyName;
        }
        
        private Container Container { get; }

        private string PartitionKeyName { get; }

        private string ETag { get; set; }

        protected override string VersionId => ETag;

        protected override async Task PutLockDataAsync()
        {
            var item = JObject.FromObject(new
            {
                Id = Uri.EscapeDataString($"Mutex-{MutexName}"),
                MutexHolder,
                Timestamp = DateTime.UtcNow
            });

            var partitionKey =
                !string.IsNullOrWhiteSpace(PartitionKeyName) ? new PartitionKey("Mutex") : PartitionKey.None;
            
            var response = await Container.CreateItemAsync(item, partitionKey);
        
            ETag = response.ETag;
        }

        protected override async Task<LockData> GetLockDataAsync()
        {
            var id = Uri.EscapeDataString($"Mutex-{MutexName}");
            var partitionKey = !string.IsNullOrWhiteSpace(PartitionKeyName) ? new PartitionKey("Mutex") : PartitionKey.None;

            var itemResponse =
                await Container.ReadItemAsync<LockData>(id, partitionKey, new ItemRequestOptions {ConsistencyLevel = ConsistencyLevel.Strong});
            if (itemResponse.StatusCode == HttpStatusCode.NotFound || itemResponse.Resource == null)
                return null;

            // sanity check which removes the record from CosmosDB in case it has incompatible structure. Only possible
            // if modified externally, but this could lead to a situation where the lock would never be acquired.
            if (itemResponse.Resource != null &&
                (itemResponse.Resource.MutexHolder == null || itemResponse.Resource.Timestamp == default))
            {
                await Container.DeleteItemAsync<LockData>(id, partitionKey);
                return null;
            }

            itemResponse.Resource.VersionId = itemResponse.ETag;
        
            return itemResponse.Resource;
        }

        protected override async Task DeleteLockDataAsync(string versionId)
        {
            var id = Uri.EscapeDataString($"Mutex-{MutexName}");
            var partitionKey = !string.IsNullOrWhiteSpace(PartitionKeyName) ? new PartitionKey("Mutex") : PartitionKey.None;

            await Container.DeleteItemAsync<LockData>(id, partitionKey, new ItemRequestOptions {IfMatchEtag = versionId});
        }
    }
}

File: azure/cosmos-db/lib/cosmos-db-mutex.ts
import { DocumentDatabaseMutex, LockData } from "@mcma/data";
import { ConsistencyLevel, Container } from "@azure/cosmos";
import { Logger } from "@mcma/core";

interface TableKey {
    id: string;
    [key: string]: any;
}

type TableItem = TableKey & LockData;

function replaceIllegalCharsInMutexName(mutexName: string): string {
    return mutexName.replace(/\//g, "_");
}

export class CosmosDbMutex extends DocumentDatabaseMutex {
    private etag: string;
    
    constructor(
        private container: Container,
        private partitionKeyName: string,
        mutexName: string,
        mutexHolder: string,
        lockTimeout: number = 300000,
        logger?: Logger
    ) {
        super(mutexName, mutexHolder, lockTimeout, logger);
    }
    
    protected get versionId(): string {
        return this.etag;
    }

    private generateTableKey(): TableKey {
        return this.partitionKeyName
            ? {
                id: replaceIllegalCharsInMutexName(this.mutexName),
                [this.partitionKeyName]: "Mutex"
            }
            : {
                id: replaceIllegalCharsInMutexName(`Mutex-${this.mutexName}`)
            };
    }

    private generateTableItem() {
        return Object.assign({}, this.generateTableKey(), {
            mutexHolder: this.mutexHolder,
            timestamp: Date.now()
        });
    }

    protected async getLockData(): Promise<LockData> {
        const item =
            this.partitionKeyName
                ? this.container.item(replaceIllegalCharsInMutexName(this.mutexName), "Mutex")
                : this.container.item(replaceIllegalCharsInMutexName(`Mutex-${this.mutexName}`));
                
        const itemResponse = await item.read<TableItem>({ consistencyLevel: ConsistencyLevel.Strong });

        if (!itemResponse.resource) {
            return undefined;
        }

        // sanity check which removes the record from CosmosDB in case it has incompatible structure. Only possible
        // if modified externally, but this could lead to a situation where the lock would never be acquired.
        if (!itemResponse.resource.mutexHolder || !itemResponse.resource.timestamp) {
            await item.delete();
            return undefined;
        }

        itemResponse.resource.versioId = itemResponse.etag;

        return itemResponse.resource;
    }

    protected async putLockData(): Promise<void> {
        const item = this.generateTableItem();
        
        const response = await this.container.items.upsert(item, {
            accessCondition: {
                type: "IfMatch",
                condition: null,
            }
        });
        
        this.etag = response.etag;
    }

    protected async deleteLockData(versionId: string): Promise<void> {
        const item =
            this.partitionKeyName
                ? this.container.item(replaceIllegalCharsInMutexName(this.mutexName), "Mutex")
                : this.container.item(replaceIllegalCharsInMutexName(`Mutex-${this.mutexName}`));

        await item.delete<TableItem>({
            accessCondition: {
                type: "IfMatch",
                condition: versionId,
            }
        });
    }
}
